app/user_test.go

```diff
@@ -653,6 +654,89 @@ func TestCreateUserWithToken(t *testing.T) {
+		)
+		grantedDomainToken := model.NewToken(
+			TOKEN_TYPE_GUEST_INVITATION,
+			model.MapToJson(map[string]string{"teamId": th.BasicTeam.Id, "email": grantedInvitationEmail, "channels": th.BasicChannel.Id}),
+		)
+		require.Nil(t, th.App.Srv.Store.Token().Save(forbiddenDomainToken))
+		require.Nil(t, th.App.Srv.Store.Token().Save(grantedDomainToken))
+		guest := model.User{
+			Email:       strings.ToLower(model.NewId()) + "+test@example.com",
+			Nickname:    "Darth Vader",
+			Username:    "vader" + model.NewId(),
+			Password:    "passwd1",
+			AuthService: "",
+		}
+		newGuest, err := th.App.CreateUserWithToken(&guest, forbiddenDomainToken)
+		require.NotNil(t, err)
+		require.Nil(t, newGuest)
+		errRegexp := regexp.MustCompile(`does not belong to an accepted domain`)
+		require.Equal(t, "does not belong to an accepted domain", errRegexp.FindString(err.Error()))
+
+		newGuest, err = th.App.CreateUserWithToken(&guest, grantedDomainToken)
+		require.Nil(t, err, "Should add user to the team. err=%v", err)
+		assert.True(t, newGuest.IsGuest())
+		require.Equal(t, grantedInvitationEmail, newGuest.Email, "The user email must be the invitation one")
+		_, err = th.App.Srv.Store.Token().GetByToken(grantedDomainToken.Token)
+		require.NotNil(t, err, "The token must be deleted after be used")
+
+		members, err := th.App.GetChannelMembersForUser(th.BasicTeam.Id, newGuest.Id)
+		require.Nil(t, err)
+		require.Len(t, *members, 1)
+		assert.Equal(t, (*members)[0].ChannelId, th.BasicChannel.Id)
+	})
+
+	t.Run("create guest having team and system email domain restrictions", func(t *testing.T) {
+		th.BasicTeam.AllowedDomains = "restricted-team.com"
+		_, err := th.App.UpdateTeam(th.BasicTeam)
+		require.Nil(t, err, "Should update the team")
+		enableGuestDomainRestricions := *th.App.Config().TeamSettings.RestrictCreationToDomains
+		defer func() {
+			th.App.UpdateConfig(func(cfg *model.Config) {
+				cfg.TeamSettings.RestrictCreationToDomains = &enableGuestDomainRestricions
+			})
+		}()
+		th.App.UpdateConfig(func(cfg *model.Config) { *cfg.TeamSettings.RestrictCreationToDomains = "restricted.com" })
+		invitationEmail := model.NewId() + "other-email@test.com"
+		token := model.NewToken(
+			TOKEN_TYPE_GUEST_INVITATION,
+			model.MapToJson(map[string]string{"teamId": th.BasicTeam.Id, "email": invitationEmail, "channels": th.BasicChannel.Id}),
+		)
+		require.Nil(t, th.App.Srv.Store.Token().Save(token))
+		guest := model.User{
+			Email:       strings.ToLower(model.NewId()) + "+test@example.com",
+			Nickname:    "Darth Vader",
+			Username:    "vader" + model.NewId(),
+			Password:    "passwd1",
+			AuthService: "",
+		}
+		newGuest, err := th.App.CreateUserWithToken(&guest, token)
+		require.Nil(t, err, "Should add user to the team. err=%v", err)
+		assert.True(t, newGuest.IsGuest())
+		require.Equal(t, invitationEmail, newGuest.Email, "The user email must be the invitation one")
```

Comment:
Same message as above. Or we can just remove it. The user will see the line number anyways. Up to you.